runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	}
+
+	err = releaseConn()
+	if err != nil {
+		return err
+	}
+
+	// replicate
+	err = d.syncBackup(ctx, tableName)
+	if err != nil {
+		return err
+	}
+	d.writeDirty = false
+	// remove old version
+	_ = d.deleteBackup(ctx, tableName, version)
+	return d.syncRead(ctx)
+}
+
+func (d *db) syncRead(ctx context.Context) error {
+	entries, err := os.ReadDir(d.writePath)
+	if err != nil {
+		return err
+	}
+
+	tableVersion := make(map[string]string)
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		// Check if there is already a table with the same version
+		writeVersion, exist, _ := d.writeTableVersion(entry.Name())
+		if !exist {
+			continue
+		}
+		tableVersion[entry.Name()] = writeVersion
+		readVersion, _, _ := d.readTableVersion(entry.Name())
+		if writeVersion == readVersion {
+			continue
+		}
+
+		d.logger.Debug("Sync: copying table", slog.String("table", entry.Name()))
+		err = copyDir(filepath.Join(d.readPath, entry.Name()), filepath.Join(d.writePath, entry.Name()))
+		if err != nil {
+			return err
+		}
+	}
+
+	handle, err := d.openDBAndAttach(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	var oldDBHandle *sqlx.DB
+	d.readMu.Lock()
+	// swap read handle
+	oldDBHandle = d.readHandle
+	d.readHandle = handle
+	d.readMu.Unlock()
+
+	// close old read handle
+	if oldDBHandle != nil {
+		err = oldDBHandle.Close()
+		if err != nil {
+			d.logger.Warn("error in closing old read handle", slog.String("error", err.Error()))
+		}
+	}
+
+	// delete data for tables/versions that have been removed from write
+	entries, err = os.ReadDir(d.readPath)
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		readVersion, ok, _ := d.readTableVersion(entry.Name())
+		if !ok {
+			// invalid table
+			_ = os.RemoveAll(filepath.Join(d.readPath, entry.Name()))
+			continue
+		}
+
+		writeVersion, ok := tableVersion[entry.Name()]
+		if !ok {
+			// table not in write
+			d.logger.Debug("Sync: removing table", slog.String("table", entry.Name()))
+			err = os.RemoveAll(filepath.Join(d.readPath, entry.Name()))
+			if err != nil {
+				return err
+			}
+			continue
+		}
+
+		if readVersion == writeVersion {
+			continue
+		}
+
+		d.logger.Debug("Sync: removing old version", slog.String("table", entry.Name()), slog.String("version", readVersion))
+		err = os.RemoveAll(filepath.Join(d.readPath, entry.Name(), readVersion))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *db) Size() int64 {
+	var paths []string
+	entries, err := os.ReadDir(d.readPath)
+	if err != nil { // ignore error
+		return 0
+	}
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		// this is to avoid counting temp tables during source ingestion
+		// in certain cases we only want to compute the size of the serving db files
+		// TODO :: remove this when removing staged table concepts
+		if strings.HasPrefix(entry.Name(), "__rill_tmp_") {
+			continue
+		}
```

Comment:
I think we should count the whole size of the DB at the given time. That's what we do now, right? 

Also, I'm not sure we can remove the staged table concept – it might be implemented differently for different OLAPs, but the concept I think is necessary? Like in cloud, we want to keep serving old data, but on local we want to drop the old data no matter if the insert is successful or not.

Reply:
No this was changed at one point based on product inputs. There were not happy with the storage metric jumping to more than 2x during certain intervals of day and wanted it be consistent with how much data is being used to serve dashboards.

Hmm. I see. I think we discussed about removing it for clickhouse so thought may be it can be removed.